testplan/runnable/base.py

```diff
@@ -393,106 +394,124 @@ def schedule(self, task=None, resource=None, uid=None, **options):
-            or self.cfg.test_lister is not None
-        ):
-            if not self.should_be_added(runnable):
-                return None
-
         local_runner = self.resources.first()
-        if resource is None:
-            resource = local_runner
+        resource = resource or local_runner
 
         if resource not in self.resources:
             raise RuntimeError(
                 'Resource "{}" does not exist.'.format(resource)
             )
 
-        # When running interactively, add all tasks to the local runner even
-        # if they were scheduled into a pool. It greatly simplifies the
-        # interactive runner if it only has to deal with the local runner.
+        # Get the real test entity and verify if it should be added
+        runnable = self._verify_test_target(target)
+        if not runnable:
+            return None
+
+        uid = runnable.uid()
+        part = (
+            getattr(runnable.cfg, "part", None)
+            if hasattr(runnable, "cfg")
+            else None
+        )
+
+        # When running interactively, add all real test entities into the local
+        # runner even if they were scheduled into a pool. It greatly simplifies
+        # the interactive runner if it only has to deal with the local runner.
         if self.cfg.interactive_port is not None:
-            if isinstance(runnable, Task):
-                runnable = runnable.materialize()
-                runnable.cfg.parent = self.cfg
-                runnable.parent = self
-            self.resources[local_runner].add(runnable, runnable.uid() or uid)
-        else:
-            self.resources[resource].add(runnable, uid)
+            if part:
+                # Modify Multitest-part instance to avoid duplicate uid
+                uid = "{} - part({}/{})".format(uid, part[0], part[1])
+                runnable.cfg._options["name"] = uid
+            if uid in self._tests:
+                raise ValueError(
+                    '{} with uid "{}" already added.'.format(target, uid)
+                )
+            self._tests[uid] = local_runner
+            self.resources[local_runner].add(runnable, uid)
+            return uid
+
+        # In batch mode uid of test entity MUST be unique unless that test
+        # entity a Multitest part instance that wrapped in a task. The uid
+        # should always be recorded in `self._tests` to avoid duplicate.
+        if uid in self._tests:
+            if not (
+                isinstance(target, Task) and part and self._tests[uid] is None
+            ):
+                raise ValueError(
+                    '{} with uid "{}" already added.'.format(target, uid)
+                )
+
+        if isinstance(target, Task):
+            if part:
+                self._tests[uid] = None  # placeholder for Multitest parts
+            if target.uid() in self._tests:
+                raise ValueError("Task {} already added.".format(target))
+            uid = target.uid()
```

Comment:
Shall we override task's uid with the uid of its content?

Reply:
overwritten